### PR TITLE
Refactor install_slurm recipe

### PR DESCRIPTION
**Pull request by NSsirena**
### Description of changes
* Explicitly list slurm dependencies
* move version definition in slurm cookbook
* make use of the newly defined munge resource
* small code clean up 
* Add tests for install_slurm recipe

### Tests
* Tests validated on all platforms on EC2.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by awslitvit** [comment]
I would also change the PR description, as you did much more than just adding tests.

File: cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Slurm attributed shared between install_slurm and configure_slurm_accounting
+default['cluster']['reserved_base_uid'] = 400
```
Comment (jdeamicis) [inline cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb]: This should be global. Please remove it and leave it under the global defaults.

File: cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+default['cluster']['slurm']['user'] = 'slurm'
+default['cluster']['slurm']['user_id'] = node['cluster']['reserved_base_uid'] + 1
```
Comment (jdeamicis) [inline cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb]: Same comment as [this one for the munge recipes](https://github.com/aws/aws-parallelcluster-cookbook/pull/1753#discussion_r1104317554). Could we please have the uid offsets managed in a central location under the global defaults or whatever other location we will be using for global parameters?

File: cookbooks/aws-parallelcluster-slurm/recipes/install_slurm.rb [hunk 3]
```diff
@@ -17,46 +17,59 @@
...
+slurm_version = '22-05-8-1'
+slurm_commit = ''
+slurm_tar_name = if slurm_commit.empty?
+                   "slurm-#{slurm_version}"
+                 else
+                   "#{slurm_commit}"
+                 end
+slurm_tarball = "#{node['cluster']['sources_dir']}/#{slurm_tar_name}.tar.gz"
+slurm_url = "https://github.com/SchedMD/slurm/archive/#{slurm_tar_name}.tar.gz"
+slurm_sha256 = '8c8f6a26a5d51e6c63773f2e02653eb724540ee8b360125c8d7732314ce737d6'
```
Comment (jdeamicis) [inline cookbooks/aws-parallelcluster-slurm/recipes/install_slurm.rb]: Why are these here and not in `attributes/slurm_attributes.rb`? Surely at least the version, commit and sha256 variables must be exposed in the attributes file. The other ones are mostly internal logic variables and can be part of the internal of the recipe/resource.

File: cookbooks/aws-parallelcluster-slurm/recipes/install_slurm.rb [hunk 4]
```diff
@@ -17,46 +17,59 @@
...
 remote_file slurm_tarball do
-  source node['cluster']['slurm']['url']
+  source slurm_url
   mode '0644'
   retries 3
   retry_delay 5
-  not_if { ::File.exist?(slurm_tarball) }
-end
+  action :create_if_missing
+end unless redhat_ubi?
```
Comment (jdeamicis) [inline cookbooks/aws-parallelcluster-slurm/recipes/install_slurm.rb]: Why would this be different for redhat? Or am I missing something about `redhat_ubi?`? This applies also to the other steps below.

File: kitchen.recipes.yml [hunk 5]
```diff
@@ -397,3 +397,22 @@ suites:
...
+        - slurm_library_shared
+    attributes:
+      dependencies:
+        - recipe:aws-parallelcluster::test_dummy
+        - recipe:aws-parallelcluster-install::directories
+        - resource:package_repos
+        - resource:install_packages
+        - recipe:aws-parallelcluster-install::python
+        - recipe:aws-parallelcluster-slurm::install_jwt
+        - recipe:aws-parallelcluster-slurm::install_pmix
```
Comment (jdeamicis) [inline kitchen.recipes.yml:422]: munge is also required for the installation of Slurm. Please add it to the dependencies.

File: kitchen.recipes.yml [hunk 6]
```diff
@@ -397,3 +397,22 @@ suites:
...
+  - name: slurm_installed
+    run_list:
+      - recipe[aws-parallelcluster::add_dependencies]
+      - recipe[aws-parallelcluster-slurm::install_slurm]
+    verifier:
+      controls:
+        - slurm_installed
+        - slurm_user_and_group_created
+        - slurm_licence_configured
+        - slurm_library_shared
```
Comment (jdeamicis) [inline kitchen.recipes.yml:413]: This is probably an extra task, but we may want to have a further test that parses the `config.log` produced by the the Slurm configure step to check that Slurm is built with all the functionalities we expect to be there.

If we don't want to address this right now, please create a task for it.

File: test/recipes/controls/aws_parallelcluster_slurm/install_slurm_spec.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+#
+# or in the "LICENSE.txt" file accompanying this file.
+# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+slurm_user = 'slurm'
+slurm_group = slurm_user
+slurm_license_path = '/opt/parallelcluster/licenses/slurm'
+
+control 'slurm_installed' do
```
Comment (jdeamicis) [inline test/recipes/controls/aws_parallelcluster_slurm/install_slurm_spec.rb:17]: We may want to expand this:
- we should be able to find the most common binaries under `/opt/slurm/bin` (e.g. `srun`,`sbatch`, `scontrol`, etc.);
- we may search for specific libraries we expect to be there (for example `accounting_storage_mysql.so` to check that the mysql plugin has been compiled);
- later on we could think of more stuff.

**Review by jdeamicis — COMMENTED** [review]
Looks good to me, I left some comments around the code for improvements/clarifications.

File: cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb (hunk 1 above)
Comment (NSsirena) [reply cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb]: Yep. 
Good catch .. I moved here for a test and forgot to remove it.
Fixing it.

File: cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb (hunk 2 above)
Comment (NSsirena) [reply cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb]: Same as above ..fixing.

File: cookbooks/aws-parallelcluster-slurm/recipes/install_slurm.rb (hunk 3 above)
Comment (NSsirena) [reply cookbooks/aws-parallelcluster-slurm/recipes/install_slurm.rb]: These aren't used elsewhere so I moved them in the smallest required scope.
I can move them back in the attributes section if you find it more readable.

File: cookbooks/aws-parallelcluster-slurm/recipes/install_slurm.rb (hunk 4 above)
Comment (NSsirena) [reply cookbooks/aws-parallelcluster-slurm/recipes/install_slurm.rb]: Yep.. unfortunately  the repositories in `redhat_ubi` don't have all the required packages.
So we agreed for this first round to skip redhat tests on docker and address this issue at a later time.

File: kitchen.recipes.yml (hunk 5 above)
Comment (NSsirena) [reply kitchen.recipes.yml:422]: I will add it for completeness, however I will need to rebase this after we merge the change that transforms the munge recipe into a resource

File: kitchen.recipes.yml (hunk 6 above)
Comment (NSsirena) [reply kitchen.recipes.yml:413]: Yes. Improving the current recipes is beyond the scope of the task...but, I agree we should add all the required checks. I'll create a task for this.

Note: config.log will be present only after the configuration phase, that still hasn't been addressed.

File: test/recipes/controls/aws_parallelcluster_slurm/install_slurm_spec.rb (hunk 7 above)
Comment (NSsirena) [reply test/recipes/controls/aws_parallelcluster_slurm/install_slurm_spec.rb:17]: I only partially agree.
I do not think we should test for the presence of the binaries, we are building the code from the sources ...so this check would actually test the `slurm makefile` and install process ..that is outside our scope.

I however agree on testing for the presence of the expected libraries since this may depend on how we configure the environment while building slurm, so something may be missing if we have some bug or regression.

I'll add this section on the NiceToHave task.

File: cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# Slurm attributes shared between install_slurm and configure_slurm_accounting
+default['cluster']['slurm']['commit'] = ''
```
Comment (eantonin) [inline cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb:2]: question, non-blocking, what's the difference between commit and sha? are both needed?

Comment (jdeamicis) [reply cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb:2]: The sha is the checksum of the tarball we download from GitHub, not the commit hash.

Comment (NSsirena) [reply cookbooks/aws-parallelcluster-slurm/attributes/slurm_attributes.rb:2]: To add a few cent, the commit can be used to pick a very specific slurm version instead of an officially released one. Like for example to test a patch or a feature that is still in development.

**Review by eantonin — APPROVED** [review]
LGTM
